### PR TITLE
feat!, perf: spanStackTraceMinDuration config option, span.stacktrace disabled by default

### PR DESCRIPTION
**Pull request by trentm**
**Collection of stacktraces for spans is now disabled by default**, because the
CPU performance overhead has been found to be too large in practice. This can be
a significant change for applications that are heavily loaded and/or produce
many spans. The relevant config option is the new `spanStackTraceMinDuration`.
Span stacktraces going away in the Kibana APM app could surprise users,
but this is not considered a "breaking" change.

Adds a **new config option `spanStackTraceMinDuration`** that replaces both
`captureSpanStackTraces` and `spanFramesMinDuration`. The latter two are now
deprecated. The new config option, `span_stack_trace_min_duration` in central
config and in other agents, is the specified cross-agent option name. It has
these differences with `spanFramesMinDuration`:
- new name
- the semantics of zero and negative values has flipped: negative means *never*
  collect span stack traces, zero means *always* collect span stack traces
  regardless of span duration
- the default unit is now milliseconds (copying the Java agent)

Other related changes:
- The performance tuning guide has been updated: The span stack trace section
  is closer to the top. I also removed most of the "APM Server communication"
  section because IME those config vars are not significant to performance.
- Normalizing of duration options has been improved to: support different
  default units, support the eventual usage of the 'us' unit for upcoming work.
- The 'transaction-and-span-overhead-realistic-size' benchmark was changed to
  use the default config, and 'transaction-and-span-overhead-realistic-size-with-stack-trace'
  was added for the same test, but always with span.stacktrace collection.

Closes: #2381
Closes: #2536
Closes: #2092

### Checklist

- [x] Implement code
- [x] Add tests
- [x] Update TypeScript typings
- [x] Update documentation
- [x] Add CHANGELOG.asciidoc entry (I'm waiting until we get a release out with current changes before adding this changelog entry -- to avoid merge pain).
- [x] Commit message follows [commit guidelines](https://github.com/elastic/apm-agent-nodejs/blob/main/CONTRIBUTING.md#commit-message-guidelines)

### CODE REVIEW

**Comment by trentm** [comment]
The argument for why this change can be done in a minor version (i.e. is not a "breaking change") is here: https://github.com/elastic/apm/pull/585#issuecomment-1016688401

That thread also includes other relevant discussion.

**Comment by trentm** [comment]
# central config test

I did a manual central config test of `span_stack_trace_min_duration`.

```js
// foo.js
const apm = require('.').start({
  apiRequestTime: '2s'
})

function makeATransAndSpan () {
  console.warn('-- interval start', )
  var t = apm.startTransaction('t')
  var s = t.startSpan('s')
  setTimeout(function () {
    s.end()
    t.end()
  }, 50)
}

makeATransAndSpan()
setInterval(makeATransAndSpan, 5000)
```

And setup mockapmserver.js to respond with a `span_stack_trace_min_duration=42ms`:

```
[2022-02-07T18:16:12.598Z]  INFO: mockapmserver/30974 on pink.local: request (req.remoteAddress=::ffff:127.0.0.1, req.remotePort=63542, req.bodyLength=0, req.body="", res.bodyLength=41)
    GET /config/v1/agents?service.name=elastic-apm-node&service.environment=development HTTP/1.1
    accept: application/json
    user-agent: apm-agent-nodejs/3.27.0 (elastic-apm-node 3.27.0)
    host: localhost:8200
    connection: keep-alive
    --
    HTTP/1.1 200 OK
    Date: Mon, 07 Feb 2022 18:16:12 GMT
    Connection: keep-alive
    Keep-Alive: timeout=5
    Transfer-Encoding: chunked

    {"span_stack_trace_min_duration": "42ms"}
```

Here we expect the *first* span (that is created before the first central config
fetch is complete) to *not* have a `span.stacktrace`. Subsequent ones *should*
have a `span.stacktrace`. This is what was observed:


```
$ ELASTIC_APM_LOG_LEVEL=debug node foo.js | ecslog
...
-- interval start
[2022-02-07T18:16:12.580Z] DEBUG (elastic-apm-node): start trace {"trans":"fe7b9f552d447a53","trace":"ded7ff91fe481b0396724439f79b1441","name":"unnamed","type":null,"subtype":null,"action":null}
...
[2022-02-07T18:16:12.581Z] DEBUG (elastic-apm-node): start span {"span":"94811a6524eab2a0","parent":"fe7b9f552d447a53","trace":"ded7ff91fe481b0396724439f79b1441","name":"s","type":null,"subtype":null,"action":null}
...
[2022-02-07T18:16:12.599Z] DEBUG (elastic-apm-node): central config received
    remoteConf: {"span_stack_trace_min_duration": "42ms"}
[2022-02-07T18:16:12.599Z]  INFO (elastic-apm-node): Central config success: updated spanStackTraceMinDuration: 0.042
[2022-02-07T18:16:12.631Z] DEBUG (elastic-apm-node): ended span {"span":"94811a6524eab2a0","parent":"fe7b9f552d447a53","trace":"ded7ff91fe481b0396724439f79b1441","name":"s","type":null,"subtype":null,"action":null}
...
[2022-02-07T18:16:12.634Z] DEBUG (elastic-apm-node): sending span {"span":"94811a6524eab2a0","parent":"fe7b9f552d447a53","trace":"ded7ff91fe481b0396724439f79b1441","name":"s","type":null}
-- interval start
[2022-02-07T18:16:17.582Z] DEBUG (elastic-apm-node): start trace {"trans":"054850add20e5b63","trace":"b8f9f03d2c101b8d262552f6a51562c3","name":"unnamed","type":null,"subtype":null,"action":null}
...
[2022-02-07T18:16:17.582Z] DEBUG (elastic-apm-node): start span {"span":"26bdd34b61c2f8a1","parent":"054850add20e5b63","trace":"b8f9f03d2c101b8d262552f6a51562c3","name":"s","type":null,"subtype":null,"action":null}
...
```

The first sent span (id=94811a6524eab2a0) correctly had no stacktrace:

```json
    {
        "span": {
            "name": "s",
            "type": "custom",
            "id": "94811a6524eab2a0",
            "transaction_id": "fe7b9f552d447a53",
            "parent_id": "fe7b9f552d447a53",
            "trace_id": "ded7ff91fe481b0396724439f79b1441",
            "subtype": null,
            "action": null,
            "timestamp": 1644257772580542,
            "duration": 50.342,
            "sync": false,
            "outcome": "success",
            "sample_rate": 1
        }
    }
```

and subsequent ones did:

```json
    {
        "span": {
            "name": "s",
            "type": "custom",
            "id": "26bdd34b61c2f8a1",
            "transaction_id": "054850add20e5b63",
            "parent_id": "054850add20e5b63",
            "trace_id": "b8f9f03d2c101b8d262552f6a51562c3",
            "subtype": null,
            "action": null,
            "timestamp": 1644257777581849,
            "duration": 50.496,
            "stacktrace": [
                {
                    "filename": "lib/instrumentation/transaction.js",
                    "lineno": 161,
                    "function": "Transaction.createSpan",
                    "library_frame": false,
                    "abs_path": "/Users/trentm/el/apm-agent-nodejs5/lib/instrumentation/transaction.js"
                },
                {
                    "filename": "lib/instrumentation/transaction.js",
                    "lineno": 120,
                    "function": "Transaction.startSpan",
                    "library_frame": false,
                    "abs_path": "/Users/trentm/el/apm-agent-nodejs5/lib/instrumentation/transaction.js"
                },
                {
                    "filename": "foo.js",
                    "lineno": 9,
                    "function": "makeATransAndSpan",
                    "library_frame": false,
                    "abs_path": "/Users/trentm/el/apm-agent-nodejs5/foo.js"
                },
                {
                    "filename": "node:internal/timers",
                    "lineno": 557,
                    "function": "listOnTimeout",
                    "library_frame": true,
                    "abs_path": "node:internal/timers"
                },
                {
                    "filename": "node:internal/timers",
                    "lineno": 500,
                    "function": "processTimers",
                    "library_frame": true,
                    "abs_path": "node:internal/timers"
                }
            ],
            "sync": false,
            "outcome": "success",
            "sample_rate": 1
        }
    }
```

**Review by astorm — APPROVED** [review]

File: test/benchmarks/006-transaction-and-span-overhead-realistic-size-with-stack-trace.js [hunk 1]
```diff
@@ -0,0 +1,50 @@
+'use strict'
```
[the remark this answers is not in the document]
Comment (trentm) [reply test/benchmarks/006-transaction-and-span-overhead-realistic-size-with-stack-trace.js:1]: Yup. I was going to add it to the dashboard if necessary.

[the remark this answers is not in the document]
Comment (trentm) [reply test/benchmarks/006-transaction-and-span-overhead-realistic-size-with-stack-trace.js:1]: The added benchmark is on the dashboard automatically.

Screen Shot 2022-02-11 at 11 39 55 AM

**Pull request merged**